### PR TITLE
Utility functions for optimizer.

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

File: ast/src/analyzed/mod.rs [hunk 1]
```diff
@@ -87,6 +89,109 @@ impl<T> Analyzed<T> {
...
+        let fixed_replacements = self
+            .constant_polys_in_source_order()
+            .iter()
+            .fold(
+                (0, BTreeMap::new()),
+                |(shift, replacements), (poly, _def)| {
+                    fold_replacement(to_remove, shift, replacements, poly.into())
+                },
+            )
+            .1;
```
Comment (lvella) [inline ast/src/analyzed/mod.rs:137]: Not sure it compiles, but to reduce repeated code, how do you like the following?
```rs
let [witness_replacements, fixed_replacements] = [
    self.committed_polys_in_source_order(),
    self.constant_polys_in_source_order()
].map(...);
```

File: ast/src/analyzed/mod.rs [hunk 2]
```diff
@@ -87,6 +89,109 @@ impl<T> Analyzed<T> {
...
+                    PolynomialType::Constant => poly.poly_id = Some(fixed_replacements[&poly_id]),
+                    PolynomialType::Intermediate => unimplemented!("Not yet implemented"),
+                }
+            }
+            ControlFlow::Continue::<()>(())
+        });
+    }
+
+    /// Adds a polynomial identity and returns the ID.
+    pub fn append_polynomial_identity(
```
Comment (lvella) [inline ast/src/analyzed/mod.rs:172]: Who is using this function?

File: ast/src/analyzed/mod.rs [hunk 3]
```diff
@@ -87,6 +89,109 @@ impl<T> Analyzed<T> {
...
+        &mut self,
+        identity: Expression<T>,
+        source: SourceRef,
+    ) -> u64 {
+        let id = self.identities.len() as u64;
+        self.identities.push(Identity {
+            id,
```
Comment (lvella) [inline ast/src/analyzed/mod.rs:179]: Why identity uses `u64` here, and uses `usize` inside `StatementIdentifier::Identity`?

File: ast/src/analyzed/mod.rs (hunk 2 above)
Comment (chriseth) [reply ast/src/analyzed/mod.rs:172]: The optimizer will use it, in the case part of a lookup can be extracted into an identity.

File: ast/src/analyzed/mod.rs (hunk 3 above)
Comment (chriseth) [reply ast/src/analyzed/mod.rs:179]: Yeah, I don't know. In some places one is more convenient than the other...

**Comment by chriseth** [comment]
Just found a bug here.

File: ast/src/analyzed/mod.rs [hunk 4]
```diff
@@ -87,6 +89,103 @@ impl<T> Analyzed<T> {
...
+                            );
+                            (shift, replacements)
+                        }
+                    },
+                )
+                .1
+        })
+        .into_iter()
+        .flatten()
+        .collect();
```
Comment (chriseth) [inline ast/src/analyzed/mod.rs:131]: Collected this into a single map.

File: ast/src/analyzed/mod.rs [hunk 5]
```diff
@@ -87,6 +89,103 @@ impl<T> Analyzed<T> {
...
+        });
+        self.source_order.retain(|s| {
+            if let StatementIdentifier::Definition(name) = s {
+                if names_to_remove.contains(name) {
+                    return true;
+                }
+            }
+            false
+        });
+        self.definitions.values_mut().for_each(|(poly, _def)| {
```
Comment (chriseth) [inline ast/src/analyzed/mod.rs:150]: Updating the IDs in the definitions themselves was missing.

**Review by lvella — APPROVED** [review]

**Pull request merged**